Authentication

RostyslavStepanchuk opened the pull request.

RostyslavStepanchuk committed:
primitive user vertical added for further security implementation

RostyslavStepanchuk committed:
JWT token implemented

RostyslavStepanchuk committed:
Implemented AuthenticationController with JWT token, AuthFilter. No tests, no refresh token

RostyslavStepanchuk committed:
Merge branch 'master' into authentication

skosinskiy commented on src/main/java/com/socialmedia/controller/external/security/UserCredentials.java:
```diff
@@ -0,0 +1,11 @@
+import lombok.Getter;
+
+@Data
+@Getter
```
adding @ Getter is unnecessary, @ Data already generate getters

RostyslavStepanchuk replied:
done

skosinskiy commented on src/main/java/com/socialmedia/model/ApplicationUser.java:
```diff
@@ -0,0 +1,40 @@
+@Data
+@Entity
+@Table(name = "users")
+@Getter
```
the same thing here, @ Data already includes getters and setters

RostyslavStepanchuk replied:
done

skosinskiy commented on src/main/java/com/socialmedia/model/ApplicationUser.java:
```diff
@@ -0,0 +1,40 @@
+public class ApplicationUser {
+
+  @Id
+  private String username;
```
it woult be better to specify column name explicitly for all fields, refactoring field may cause change hibernate mappings, but db table will know nothing about this

RostyslavStepanchuk replied:
done

skosinskiy commented on src/main/java/com/socialmedia/model/ApplicationUser.java:
```diff
@@ -0,0 +1,40 @@
+  @Id
+  private String username;
+  @ToString.Exclude
+  @Setter
```
not neccessary

RostyslavStepanchuk replied:
fixed

skosinskiy commented on src/main/java/com/socialmedia/controller/ResponseEntityProvider.java:
```diff
@@ -0,0 +1,24 @@
+
+import static org.springframework.http.HttpStatus.NOT_FOUND;
+
+public interface ResponseEntityProvider<T> {
```
I guess adding generic to this interface may cause troubles in situations, when you have several methods in controller and have to return different types of dto. Remove generic type from interface level and add it only to method level

RostyslavStepanchuk replied:
done

skosinskiy commented on src/main/java/com/socialmedia/controller/AuthenticationController.java:
```diff
@@ -0,0 +1,28 @@
+  }
+
+  @PostMapping("/access-token")
+  public ResponseEntity<String> getAccessJWT(@RequestBody UserCredentials credentials) {
```
It would be better to return dto object with 1 field, because in future this method will be returning refresh-token and it's expire time as welll as access token

RostyslavStepanchuk replied:
done

skosinskiy commented on src/main/java/com/socialmedia/repository/UserRepository.java:
```diff
@@ -0,0 +1,10 @@
+import java.util.Optional;
+
+public interface UserRepository extends JpaRepository<ApplicationUser, String> {
+  Optional<ApplicationUser> findByUsername(String  username);
```
Jpa repository already contains method findById. As username is id for ApplicationUser entity, this method may be removed and instead you should use findById method

RostyslavStepanchuk replied:
done

skosinskiy commented on src/main/java/com/socialmedia/security/JWTAuthenticationFilter.java:
```diff
@@ -0,0 +1,57 @@
+
+    if (header == null || !header.startsWith(TOKEN_PREFIX)) {
+      chain.doFilter(req, res);
+      return;
```
I don't like this empty return... doFilterInternal method is void, you may set authentication for reversed condition - header !=null && header.startsWith(TOKEN_PREFIX)

RostyslavStepanchuk replied:
done

skosinskiy commented on src/main/java/com/socialmedia/security/JWTAuthenticationFilter.java:
```diff
@@ -0,0 +1,57 @@
+
+    String header = req.getHeader(HEADER_STRING);
+
+    if (header == null || !header.startsWith(TOKEN_PREFIX)) {
```
for better readability you may put this condition to separate method, for example "isJwtTokenPresent"

RostyslavStepanchuk replied:
done

skosinskiy commented on src/main/java/com/socialmedia/security/JWTAuthenticationFilter.java:
```diff
@@ -0,0 +1,57 @@
+    chain.doFilter(req, res);
+  }
+
+  private UsernamePasswordAuthenticationToken getAuthentication(HttpServletRequest req) {
```
this method may be moved to AuthenticationService

RostyslavStepanchuk replied:
done

RostyslavStepanchuk committed:
All comments fixed

RostyslavStepanchuk committed:
master merged into authentication

skosinskiy commented on src/main/java/com/socialmedia/service/UserServiceImpl.java:
```diff
@@ -0,0 +1,30 @@
+  @Override
+  public Optional<ApplicationUser> getCurrentUser() {
+    // TODO this is a draft method just to get some user returned
+    return Optional.of(userRepository.findAll().get(0));
```
Almost done! Just left to fix this todo and return real user. Reference - https://www.baeldung.com/get-user-in-spring-security

skosinskiy commented on src/main/java/com/socialmedia/controller/UserController.java:
```diff
@@ -0,0 +1,49 @@
+
+  @PostMapping("/sign-up")
+  public ResponseEntity<ApplicationUser> signUp(@RequestBody ApplicationUser user) {
+    user.setPassword(bcryptPasswordEncoder.encode(user.getPassword()));
```
it would be better to do this in service layer

skosinskiy commented on src/main/java/com/socialmedia/controller/UserController.java:
```diff
@@ -0,0 +1,49 @@
+  private UserService userService;
+  private BCryptPasswordEncoder bcryptPasswordEncoder;
+
+  public UserController(UserService userService) {
```
remove duplicate constructor

RostyslavStepanchuk committed:
updated comments:
duplicated constructor removed
implemented currentUser method
moved password encryption to service layer

RostyslavStepanchuk committed:
trailing line removed
